--- a/meson.build.orig	2021-02-17 14:55:28.000000000 -0800
+++ b/meson.build	2021-02-17 14:56:13.000000000 -0800
@@ -22,7 +22,7 @@
   'mesa',
   ['c', 'cpp'],
   version : run_command(
-    [find_program('python', 'python3'), 'bin/meson_get_version.py']
+    [find_program('python', '@@PYTHON3@@'), 'bin/meson_get_version.py']
   ).stdout(),
   license : 'MIT',
   meson_version : '>= 0.52',
@@ -869,7 +869,7 @@
   pre_args += '-DHAVE_HAIKU_PLATFORM'
 endif
 
-prog_python = import('python').find_installation('python3')
+prog_python = import('python').find_installation('@@PYTHON3@@')
 has_mako = run_command(
   prog_python, '-c',
   '''
